Zend/Optimizer/zend_optimizer.c

```diff
@@ -831,6 +831,49 @@ zend_class_entry *zend_optimizer_get_class_entry_from_op1(
 	return NULL;
 }
 
+const zend_class_constant *zend_fetch_class_const_info(
+	const zend_script *script, const zend_op_array *op_array, const zend_op *opline, bool *is_prototype) {
+	const zend_class_entry *ce = NULL;
+	bool is_static_reference = false;
+
+	if (!opline || !op_array || !script || opline->op2_type != IS_CONST || Z_TYPE_P(CRT_CONSTANT(opline->op2)) != IS_STRING) {
+		return NULL;
+	}
+	if (opline->op1_type == IS_CONST) {
+		zval *op1 = CRT_CONSTANT(opline->op1);
+		if (Z_TYPE_P(op1) == IS_STRING) {
+			ce = zend_optimizer_get_class_entry(script, op_array, Z_STR_P(op1 + 1));
+		}
+	} else if (opline->op1_type == IS_UNUSED && op_array->scope && !(op_array->scope->ce_flags & ZEND_ACC_TRAIT)) {
```

Comment:
I believe, in theory, if a linked class with traits was optimized (which isn't currently the case), this should still work fine because `op_array->scope` should already be set to the correct class. In that case, we know where to look for the class constant (i.e. in `op_array->scope`, the linked class).

As of now, the two checks should be equivalent, if I'm not mistaken.

Reply:
I'm also fine with also excluding op_arrays with `ZEND_ACC_TRAIT_CLONE` just for the sake of defensiveness. I could find (at least) one instance where both flags are checked: https://github.com/php/php-src/blob/47ae33ae1173abb3dbc05c71bb02020fed2f7f22/Zend/Optimizer/zend_optimizer.c#L938

Reply:
feel free to add both